### PR TITLE
Adjust Book examples to the new deserialization API

### PR DESCRIPTION
**Pull request by wprzytula**
This PR adjusts examples in the Book to the new API, so that the tests compile and pass. This should be merged quickly to bring back the passing Book check to the CI.

Further adjustments of the Book to the new API, including more descriptive tips and a migration guide, are left as a follow-up.

Refs: #462
Depends on: #1057 

## Pre-review checklist

- [x] I have split my patch into logically separate commits.
- [x] All commit messages clearly explain what they change and why.
- ~~[ ] I added relevant tests for new features and bug fixes.~~
- [x] All commits compile, pass static checks and pass test.
- [x] PR description sums up the changes and reasons why they should be introduced.
- ~~[ ] I have provided docstrings for the public items that I want to introduce.~~
- [x] I have adjusted the documentation in `./docs/source/`.
- ~~[ ] I added appropriate `Fixes:` annotations to PR description.~~

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -21,9 +21,9 @@ let uri = "127.0.0.1:9042";
...
 let session: Session = SessionBuilder::new().known_node(uri).build().await?;
 
-let raw_iter = session.query_iter("SELECT a, b, c FROM ks.t", &[]).await?;
-let mut iter = raw_iter.into_typed::<(i32, i32, String)>();
-while let Some((a, b, c)) = iter.try_next().await? {
+let query_pager = session.query_iter("SELECT a, b, c FROM ks.t", &[]).await?;
+let mut stream = query_pager.rows_stream::<(i32, i32, String)>()?;
+while let Some((a, b, c)) = stream.try_next().await? {
     println!("a, b, c: {}, {}, {}", a, b, c);
 }
```
Comment (Lorak-mmk) [inline README.md:28]: Why not &str?

File: docs/source/data-types/blob.md [hunk 2]
```diff
@@ -18,8 +18,10 @@ session
...
 
 // Read blobs from the table
-let mut iter = session.query_iter("SELECT a FROM keyspace.table", &[]).await?.into_typed::<(Vec<u8>,)>();
-while let Some((blob_value,)) = iter.try_next().await? {
+let mut stream = session.query_iter("SELECT a FROM keyspace.table", &[])
+    .await?
+    .rows_stream::<(Vec<u8>,)>()?;
+while let Some((blob_value,)) = stream.try_next().await? {
     println!("{:?}", blob_value);
 }
```
Comment (Lorak-mmk) [inline docs/source/data-types/blob.md:26]: Why Vec and not slice?

File: docs/source/data-types/collections.md [hunk 3]
```diff
@@ -18,8 +18,10 @@ session
...
 // Read a list of ints from the table
-let mut iter = session.query_iter("SELECT a FROM keyspace.table", &[]).await?.into_typed::<(Vec<i32>,)>();
-while let Some((list_value,)) = iter.try_next().await? {
+let mut stream = session.query_iter("SELECT a FROM keyspace.table", &[])
+    .await?
+    .rows_stream::<(Vec<i32>,)>()?;
+while let Some((list_value,)) = stream.try_next().await? {
     println!("{:?}", list_value);
 }
 # Ok(())
```
Comment (Lorak-mmk) [inline docs/source/data-types/collections.md:27]: ditto

File: docs/source/data-types/collections.md [hunk 4]
```diff
@@ -44,10 +46,10 @@ session
...
 // Read a set of ints from the table
-let mut iter = session.query_iter("SELECT a FROM keyspace.table", &[])
+let mut stream = session.query_iter("SELECT a FROM keyspace.table", &[])
     .await?
-    .into_typed::<(Vec<i32>,)>();
-while let Some((set_value,)) = iter.try_next().await? {
+    .rows_stream::<(Vec<i32>,)>()?;
+while let Some((set_value,)) = stream.try_next().await? {
     println!("{:?}", set_value);
 }
```
Comment (Lorak-mmk) [inline docs/source/data-types/collections.md:54]: ditto

File: docs/source/data-types/text.md [hunk 5]
```diff
@@ -24,7 +24,7 @@ session
 // Read ascii/text/varchar from the table
 let mut iter = session.query_iter("SELECT a FROM keyspace.table", &[])
     .await?
-    .into_typed::<(String,)>();
+    .rows_stream::<(String,)>()?;
 while let Some((text_value,)) = iter.try_next().await? {
     println!("{}", text_value);
 }
```
Comment (Lorak-mmk) [inline docs/source/data-types/text.md:30]: ditto

File: docs/source/data-types/tuple.md [hunk 6]
```diff
@@ -19,7 +19,7 @@ session
 // Read a tuple of int and string from the table
 let mut iter = session.query_iter("SELECT a FROM keyspace.table", &[])
     .await?
-    .into_typed::<((i32, String),)>();
+    .rows_stream::<((i32, String),)>()?;
 while let Some((tuple_value,)) = iter.try_next().await? {
     let int_value: i32 = tuple_value.0;
     let string_value: String = tuple_value.1;
```
Comment (Lorak-mmk) [inline docs/source/data-types/tuple.md:25]: ditto

File: docs/source/queries/result.md [hunk 7]
```diff
@@ -116,8 +73,11 @@ To properly handle `NULL` values parse column as an `Option<>`:
...
-if let Some(rows) = session.query_unpaged("SELECT a, b from ks.tab", &[]).await?.rows {
-    for row in rows.into_typed::<(i32, Option<String>)>() {
+if let Some(rows_result) = session.query_unpaged("SELECT a, b from ks.tab", &[])
+    .await?
+    .into_rows_result()?
+{
+    for row in rows_result.rows::<(i32, Option<String>)>()? {
         let (int_value, str_or_null): (i32, Option<String>) = row?;
     }
 }
```
Comment (Lorak-mmk) [inline docs/source/queries/result.md:83]: ditto

**Review by Lorak-mmk — CHANGES_REQUESTED** [review]
Looks good - all of my comments are regarding one issue: that we should use borrowed deserialization where possible.

**Comment by Lorak-mmk** [comment]
Oh wait, most of those is probably because we have no lending stream, right? 
In that case please:
- Switch those that don't use stream to borrowed form
- Introduce closure API and switch over the rest (or at least those that can be switched).

File: docs/source/queries/result.md [hunk 8]
```diff
@@ -19,91 +19,48 @@ return a `QueryResult` with rows represented as `Option<Vec<Row>>`.
...
-* `maybe_first_row_typed::<RowT>` - returns `Option<RowT>` containing first row from the result
-* `first_row_typed::<RowT>` - same as `maybe_first_row`, but fails without the first row
-* `single_row_typed::<RowT>` - same as `first_row`, but fails when there is more than one row
-* `result_not_rows()` - ensures that query response was not `rows`, helps avoid bugs
+* `rows::<RowT>()` - returns the rows parsed as the given type
+* `maybe_first_row_typed::<RowT>` - returns the first received row or `None` if there are no rows
+* `first_row::<RowT>` - returns the first received row; fails if there are no rows
+* `single_row::<RowT>` - same as `first_row`, but fails when there is more than one row
 
+Additionally, [`QueryResult`](https://docs.rs/scylla/latest/scylla/transport/query_result/struct.QueryResult.html) has a method `result_not_rows()`, which ensures that query response was not `rows` and thus helps avoid bugs.
```
Comment (muzarski) [inline docs/source/queries/result.md:31]: Maybe we should mention how to obtain `QueryRowsResult` from `QueryResult` as it's not exactly clear after reading this section (without looking at examples).

**Review by muzarski — CHANGES_REQUESTED** [review]

File: README.md (hunk 1 above)
Comment (wprzytula) [reply README.md:28]: `Stream` does not permit borrowed types.

File: docs/source/data-types/blob.md (hunk 2 above)
Comment (wprzytula) [reply docs/source/data-types/blob.md:26]: ditto

**Comment by wprzytula** [comment]
> Introduce closure API and switch over the rest (or at least those that can be switched).

I'm not convinced we have to introduce the closure API in this release. After all, it's going to be a new API, not breaking anything. OTOH other APIs introduced in the deserialization refactor aren't entirely new APIs; they are rather improvements over old APIs (QueryResult, TypedRowStream etc.).
What we have so far in 0.15 is a major upgrade already, reducing a lot of needless allocations (in form of `Rows`, `Row` and `CqlValue`). I fully support deserialization of borrowed types, but the new framework already can do that - and adding the API for that in the yet next release will be seamless for the users.

**Review by muzarski — APPROVED** [review]

**Review by muzarski — APPROVED** [review]

**Review by Lorak-mmk — APPROVED** [review]

**Pull request merged**